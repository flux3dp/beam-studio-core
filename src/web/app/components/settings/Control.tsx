/* eslint-disable react/require-default-props */
import * as React from 'react';

import browser from 'implementations/browser';

interface Props {
  id?: string,
  label: string,
  url?: string,
  warningText?: string,
  children: JSX.Element | JSX.Element[],
}

const Controls = ({
  id = '',
  label,
  url = '',
  warningText = null,
  children,
}: Props): JSX.Element => {
  const style = { width: 'calc(100% / 10 * 3 - 10px)' };
  const innerHtml = { __html: label };

  const warningIcon = () => {
    if (warningText) {
      return (<img src="img/warning.svg" title={warningText} />);
    }
    return null;
  };

  const renderIcon = () => {
    if (url) {
      return (
        <span className="info-icon-small">
          <img src="img/info.svg" onClick={() => { browser.open(url); }} />
        </span>
      );
    }
    return null;
  };

  return (
    <div id={id} className="row-fluid">
      <div className="span3 no-left-margin" style={style}>
        <label
          className="font2"
          // eslint-disable-next-line react/no-danger
          dangerouslySetInnerHTML={innerHtml}
        />
        {renderIcon()}
      </div>
      <div className="span8 font3">
        {children}
        {warningIcon()}
      </div>
    </div>
  );
};

export default Controls;
