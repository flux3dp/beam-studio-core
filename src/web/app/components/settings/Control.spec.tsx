import * as React from 'react';
import toJson from 'enzyme-to-json';
import { mount, shallow } from 'enzyme';
import { render } from '@testing-library/react';

const open = jest.fn();
jest.mock('implementations/browser', () => ({
  open,
}));

const useIsMobile = jest.fn();
jest.mock('helpers/system-helper', () => ({
  useIsMobile: () => useIsMobile(),
}));

// eslint-disable-next-line import/first
import Control from './Control';

describe('test Control', () => {
  it('should render correctly', () => {
    expect(
      toJson(
        shallow(
          <Control label="Flux">
            <div>Hello World</div>
          </Control>
        )
      )
    ).toMatchSnapshot();

    const wrapper = mount(
      <Control label="Flux" url="https://www.flux3dp.com" warningText="Warning!!">
        <div>Hello World</div>
      </Control>
    );
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find('img').at(0).simulate('click');
    expect(open).toHaveBeenCalledTimes(1);
    expect(open).toHaveBeenNthCalledWith(1, 'https://www.flux3dp.com');
  });

  it('should render correctly in mobile', () => {
    useIsMobile.mockReturnValue(true);
    const { container } = render(
      <Control label="Flux">
        <div>Hello World</div>
      </Control>
    );
    expect(container).toMatchSnapshot();
  });
});
