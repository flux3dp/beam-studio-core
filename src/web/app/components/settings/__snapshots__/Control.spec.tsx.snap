// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Control should render correctly 1`] = `
<div
  className="row-fluid control"
  id=""
>
  <div
    className="no-left-margin span3 label"
  >
    <label
      className="font2"
      dangerouslySetInnerHTML={
        {
          "__html": "Flux",
        }
      }
    />
  </div>
  <div
    className="font3 span8"
  >
    <div>
      Hello World
    </div>
  </div>
</div>
`;

exports[`test Control should render correctly 2`] = `
<Controls
  label="Flux"
  url="https://www.flux3dp.com"
  warningText="Warning!!"
>
  <div
    className="row-fluid control"
    id=""
  >
    <div
      className="no-left-margin span3 label"
    >
      <label
        className="font2"
        dangerouslySetInnerHTML={
          {
            "__html": "Flux",
          }
        }
      />
      <span
        className="info-icon-small"
      >
        <img
          onClick={[Function]}
          src="img/info.svg"
        />
      </span>
    </div>
    <div
      className="font3 span8"
    >
      <div>
        Hello World
      </div>
      <img
        src="img/warning.svg"
        title="Warning!!"
      />
    </div>
  </div>
</Controls>
`;

exports[`test Control should render correctly in mobile 1`] = `
<div>
  <div
    class="row-fluid control"
    id=""
  >
    <div
      class="no-left-margin"
    >
      <label
        class="font2"
      >
        Flux
      </label>
    </div>
    <div
      class="font3"
    >
      <div>
        Hello World
      </div>
    </div>
  </div>
</div>
`;
